library/core/src/num/error.rs

```diff
@@ -77,51 +77,47 @@ pub struct ParseIntError {
 /// # Example
 ///
 /// ```
-/// #![feature(int_error_matching)]
-///
 /// # fn main() {
 /// if let Err(e) = i32::from_str_radix("a12", 10) {
 ///     println!("Failed conversion to i32: {:?}", e.kind());
 /// }
 /// # }
 /// ```
-#[unstable(
-    feature = "int_error_matching",
-    reason = "it can be useful to match errors when making error messages \
-              for integer parsing",
-    issue = "22639"
-)]
+#[stable(feature = "int_error_matching", since = "1.47.0")]
 #[derive(Debug, Clone, PartialEq, Eq)]
 #[non_exhaustive]
 pub enum IntErrorKind {
     /// Value being parsed is empty.
     ///
     /// Among other causes, this variant will be constructed when parsing an empty string.
+    #[stable(feature = "int_error_matching", since = "1.47.0")]
     Empty,
-    /// Contains an invalid digit.
+    /// Contains an digit invalid in its context.
     ///
     /// Among other causes, this variant will be constructed when parsing a string that
-    /// contains a letter.
-    InvalidDigit,
+    /// contains a non-asci char.
+    ///
+    /// This variant is also constructed when a `+` or `-` is misplaced within a sting
+    /// either on its own or in the middle of a number.
+    #[stable(feature = "int_error_matching", since = "1.47.0")]
+    InvalidDigit(#[stable(feature = "int_error_matching", since = "1.47.0")] char),
```

Comment:
```suggestion
    InvalidDigit,
```

I think if we're going to keep the enum open here we should try limit it to simple variants. That would mean we don't know which `char` we encountered that caused us to return this variant, but in general I think just returning a `char` isn't quite enough for consumers to render meaningful errors. They'd probably also want to know the position in the string that digit appears at.

So I think removing the `char` from this variant will avoid a bit of noise in match arms without losing too much.